File: ml-benchmark/fusion/tupling/page8.f
Inductive Nat = zero Unit | succ Nat;

pred = \n: Nat.
  match n with
    zero _ -> n
  | succ m -> m
  end;

fib = fix (
  \f: Nat -> Int. \n: Nat.
  match n with
    zero _ -> 0
  | succ (zero _) -> 1
  | succ (succ m) ->
    + (f m) (f (succ m))
  end
);

repr = fix (
  \f: Nat -> Compress Nat. \n: Nat.
  match n with
    zero _ -> n
  | succ m -> succ (f m)
  end
);

main = \n: Nat. fib (repr n);

File: ml-benchmark/fusion/tupling/page8.ml
type nat = Zero of unit | Succ of nat

let pred n =
  match n with
  | Zero _ -> n
  | Succ m -> m

let rec fib n =
  match n with
  | Zero _ -> 0
  | Succ m ->
      match m with
      | Zero _ -> 1
      | Succ m2 -> fib m2 + fib (Succ m2)

let rec repr n =
  match n with
  | Zero _ -> n
  | Succ m -> Succ (repr m)

let program n = fib (repr n)